src/container/LinkedCellContainer.cpp

```diff
@@ -0,0 +1,390 @@
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int y = -1; y <= nCells[1]; ++y) {
+                    haloIndices.push_back(get1DIndex({-1, y, z}));
+                }
+            }
+            break;
+        case RIGHT:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int y = -1; y <= nCells[1]; ++y) {
+                    haloIndices.push_back(get1DIndex({nCells[0], y, z}));
+                }
+            }
+            break;
+        case DOWN:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, -1, z}));
+                }
+            }
+            break;
+        case UP:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, nCells[1], z}));
+                }
+            }
+            break;
+        case BACK:
+            for (int y = -1; y <= nCells[1] ; ++y) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, y, -1}));
+                }
+            }
+            break;
+        case FRONT:
+            for (int y = -1; y <= nCells[1] ; ++y) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, y, nCells[2]}));
+                }
+            }
+    }
+    return haloIndices;
+}
+
+void LinkedCellContainer::removeFromHalo(Direction direction) {
+    for(int i : getAllHaloIndices(direction)) {
+        for (int p: cells[i].getParticleIndices()) {
+            (*particles)[p].removeFromDomain();
+            cells[i].removeIndex(p);
+        }
+    }
+}
+
+void LinkedCellContainer::updateHalo(Direction direction, BoundaryCondition boundaryCondition) {
+    for(int i : getAllHaloIndices(direction)) {
+        for (int p : cells[i].getParticleIndices()) {
+            switch (boundaryCondition) {
+                case OUTFLOW:
+                    (*particles)[p].removeFromDomain();
+                    cells[i].removeIndex(p);
+                    break;
+                case REFLECTING:
+                    // change sign of the corresponding velocity component if the particle is flying away.
+                    if ((*particles)[p].getV()[direction / 2] * (direction % 2 - 0.5) > 0) {
+                        std::array<double, 3> vel = (*particles)[p].getV();
+                        vel[direction / 2] *= -1;
+                        (*particles)[p].setV(vel);
+                    }
+                    break;
+                case PERIODIC:
+                    cells[i].removeIndex(p);
+                    std::array<double, 3> pos = (*particles)[p].getX();
+                    pos[direction / 2] = pos[direction / 2] + pow(-1, direction % 2) * domainSize[direction / 2];
+                    (*particles)[p].setX(pos);
```

Comment:
Can we generally assume that the boundary conditions in the same direction are always the same (e.g. the left and the right boundary always have the same boundary condition)? because it does not really make sense to have a periodic boundary on the left hand side and something else on the right hand side.

Reply:
It is only always true for periodic boundary, not for others. For others, you have to consider separately for each boundary. So you can indeed have cases where left is reflective while right is outflow.